File: iterai-js/src/storage.ts
import { getConfig } from "./config.js";

const logger = {
  debug: (msg: string) => console.debug(`[DEBUG] storage: ${msg}`),
};

export interface GraphData {
  nodes: Record<string, any>;
  edges: Array<{ from: string; to: string }>;
}

/**
 * Storage abstraction for IterAI
 * Uses localStorage in browser, or in-memory fallback
 */
export class Storage {
  public path: string;
  private storageKey: string;
  private useLocalStorage: boolean;
  private memoryStore: Map<string, string>;

  constructor(storagePath?: string) {
    const finalPath =
      storagePath || getConfig().get("storage.path", "iterai-storage");
    this.path = finalPath;
    this.storageKey = `iterai:${this.path}`;
    this.useLocalStorage = typeof localStorage !== "undefined";
    this.memoryStore = new Map();
    logger.debug(`Storage initialized: ${this.path}`);
  }

  private getItem(key: string): string | null {
    if (this.useLocalStorage) {
      return localStorage.getItem(key);
    }
    return this.memoryStore.get(key) || null;
  }

  private setItem(key: string, value: string): void {
    if (this.useLocalStorage) {
      localStorage.setItem(key, value);
    } else {
      this.memoryStore.set(key, value);
    }
  }

  saveGraph(graphData: GraphData): void {
    const key = `${this.storageKey}:graph`;
    this.setItem(key, JSON.stringify(graphData));
  }

  loadGraph(): GraphData {
    const key = `${this.storageKey}:graph`;
    const data = this.getItem(key);
    if (!data) {
      return { nodes: {}, edges: [] };
    }
    try {
      return JSON.parse(data);
    } catch {
      return { nodes: {}, edges: [] };
    }
  }

  saveNode(nodeId: string, nodeData: any): void {
    const key = `${this.storageKey}:node:${nodeId}`;
    this.setItem(key, JSON.stringify(nodeData));
  }

  loadNode(nodeId: string): any | null {
    const key = `${this.storageKey}:node:${nodeId}`;
    const data = this.getItem(key);
    if (!data) return null;
    try {
      return JSON.parse(data);
    } catch {
      return null;
    }
  }

  nodeExists(nodeId: string): boolean {
    const key = `${this.storageKey}:node:${nodeId}`;
    return this.getItem(key) !== null;
  }

  deleteNode(nodeId: string): void {
    const key = `${this.storageKey}:node:${nodeId}`;
    if (this.useLocalStorage) {
      localStorage.removeItem(key);
    } else {
      this.memoryStore.delete(key);
    }
  }

  clear(): void {
    if (this.useLocalStorage) {
      const keysToDelete: string[] = [];
      for (let i = 0; i < localStorage.length; i++) {
        const key = localStorage.key(i);
        if (key && key.startsWith(this.storageKey)) {
          keysToDelete.push(key);
        }
      }
      keysToDelete.forEach((k) => localStorage.removeItem(k));
    } else {
      this.memoryStore.clear();
    }
  }
}

File: iterai-py/iterai/storage.py
import json
import logging
from pathlib import Path
from uuid import UUID

from .config import get_config

logger = logging.getLogger(__name__)


class Storage:
    def __init__(self, storage_path: str | None = None):
        if storage_path is None:
            storage_path = get_config().get("storage.path", "~/.config/iterai")
        self.path = Path(storage_path).expanduser()
        self.path.mkdir(parents=True, exist_ok=True)
        (self.path / "nodes").mkdir(exist_ok=True)

    def save_graph(self, graph_data: dict):
        graph_file = self.path / "graph.json"
        with open(graph_file, "w") as f:
            json.dump(graph_data, f, indent=2)

    def load_graph(self) -> dict:
        graph_file = self.path / "graph.json"
        if not graph_file.exists():
            return {"nodes": {}, "edges": []}
        with open(graph_file, "r") as f:
            return json.load(f)

    def node_exists(self, node_id: UUID) -> bool:
        return (self.path / "nodes" / str(node_id)).exists()
